fusil/doc/configuration.rst
+++++++++++++++++++
Fusil configuration
+++++++++++++++++++

You can configure Fusil using a fusil.conf file in your configuration directory
($XDG_CONFIG_HOME environment variable or ~/.config/). Template file: ::

    ###############################################################
    # General Fusil options
    ###############################################################
    [fusil]

    # Maximum number of session (0=unlimited)
    session = 0

    # Maximum number of success before exit (0=unlimited)
    success = 1

    # Minimum score for a successful session
    success_score = 0.50

    # Maximum score for a session error
    error_score = -0.50

    # Maximum memory in bytes
    max_memory = 104857600

    # (Normal) Maximum system load
    normal_calm_load = 0.50

    # (Normal) Seconds to sleep until system load is low
    normal_calm_sleep = 0.5

    # (Slow) Maximum system load
    slow_calm_load = 0.30

    # (Slow) Seconds to sleep until system load is low
    slow_calm_sleep = 3.0

    # xhost program path (change X11 permissions)
    xhost_program = xhost

    ###############################################################
    # Debugger used to trace child processes
    ###############################################################
    [debugger]

    # Use the debugger?
    use_debugger = True

    # Enable trace forks option
    trace_forks = True


    ###############################################################
    # Child processes options
    ###############################################################
    [process]

    # Dump core on crash
    core_dump = True

    # Maximum user process (RLIMIT_NPROC)
    max_user_process = 10

    # Default maximum memory in bytes
    max_memory = 104857600

    # Change the user (setuid)
    user = fusil

    # Change the group (setgid)
    group = fusil

    # Use a probe to watch CPU activity
    use_cpu_probe = True



fusil/ChangeLog
Changelog
=========

Fusil 1.2.2
-----------

 * Create fusil-gimp
 * Remove charset from WriteCode: use builtin open() instead codecs.open()
   because files created by open() are much faster
 * Optimize FileWatch: don't recompile patterns at each session
 * fusil now depends on python-ptrace 0.6
 * Don't use close_fds argument of subprocess.Popen() on Windows
 * Fix configuration reader: normal_calm_load, normal_calm_sleep,
   slow_calm_load, slow_calm_sleep keys global options are float, not integer

Fusil 1.2.1 (2009-02-06)
------------------------

 * Fix mangle agent of the Image Magick fuzzer
 * Fix AttachProcessPID() probe: stop the probe at process exit

Fusil 1.2 (2009-02-04)
----------------------

User visible changes:

 * Fusil now requires Python 2.5
 * Documentation: write an index (index.rst) and an user guide (usage.rst)
 * Replay script: copy HOME environment for GDB and catch setuid() error
 * fusil-firefox: support more file formats (bmp, gif, ico, png, svg), create
   --test command line option, write the HTML page into index.html file
 * fusil-python: write errors to stderr (instead of stdout) to avoid unicode
   error (especially with Python3)
 * FileWatch: rename the session with "long_output" if the program wrote more
   than max_nbline lines
 * fusil-python: blacklist posix.fork() to avoid false positive
 * If the process is killed by a signal, rename the session using the
   signal name (already worked if the debugger was disabled)

Developer changes:

 * MangleAgent supports multiple input files
 * Create DummyMangle: agent with MangleFile API but don't touch file content
   to test the fuzzer
 * Network: close() method of NetworkClient and ServerClient use
   shutdown(SHUT_RDWR)
 * NetworkServer uses a backlog of 5 clients for socket.listen() (instead of 1)

Bugfixes:

 * Fix Directory.rmtree() and replay script for Python 3.0
 * Fix ServerClient.sendBytes(): use socket.send() result to get the next
   data offset

Fusil 1.1 (2008-10-22)
----------------------

User visible changes:
 * replay.py: ask confirmation if the fuzzer will not be running under a
   different user or as root
 * Even with --force-unsafe, show safety warning if the fuzzer is
   running as the root user
 * Close files for child processes (close_fds=True)
 * Fix directory.rmtree() for Python 3.0 final

Developer changes:
 * Create IntegerRangeGenerator in fusil.unicode_generator
 * Create EnvVarIntegerRange in fusil.process.env
 * Create fusil-wizzard fuzzer
 * Write timestamp in session.log
 * Add session() method to ProjectAgent
 * Add NAME attribute to a fuzzer, reused to choose the project directory name

Bugfixes:
 * Fix Debugger.processSignal(): use the process agent to send the message
   (session_rename) since the debugger agent may be disabled
 * Fix replay.py: quote gdb arguments escape quote and antislash characters
   (eg. "text=\"Hello\\n\".")
 * replay.py uses /dev/null for stdin as Fusil does
 * FileWatch: open file in binary mode to use bytes in Python3

Fusil 1.0 final (2008-09-13)
----------------------------

Visible changes:

 * Create fusil-zzuf fuzzer (use the zzuf library)
 * Create fusil-vlc fuzzer (VLC media player)
 * For each session, generate a Python script (replay.py) to replay the
   session. The script can run the target in gdb, valgrind or gdb.py
   (python-ptrace debugger), with many options (--user, --limit, etc.)
 * Create --force-unsafe option, like --unsafe without the confirmation
 * CreateProcess is now a probe (with a score): if the debugger catchs a
   fatal signal, the session stops
 * Always use a null device as stdin for child processes to avoid blocking the
   fuzzer if the process reads stdin (eg. call getchar())
 * Write the created process identifier in the logs

Developer:

 * Create EnvVarIntegerRange: environment variable with an integer value
   in a fixed range
 * Changes to get a minimal Windows support: disable "change user/group"
   feature on Windows; remove log file before removing the project directory;
   use ":NUL" instead of /dev/null for null input/output
 * On setupProject() error, make sure that the project is cleaned
 * Close stdout files (input and output) at process exit (fix needed
   by Windows)
 * Rename long2raw() to uint2bytes(), and bytes2long() to bytes2uint()
 * Normalize score that make sure that a probe score is in range [-1; +1]
   and so that score*weight is in range[-weight; +weight]
 * CodeC: remove method lines(), writeCode() is renamed writeIntoFile(),
   use unicode strings (instead of byte strings)
 * Remove StdoutFile class, code merged in CreateProcess



fusil/fusil/config.py
from ConfigParser import RawConfigParser, NoSectionError, NoOptionError
from os.path import exists as path_exists, join as path_join
from os import getenv

class ConfigError(Exception):
    pass

class FusilConfig:
    def __init__(self):
        self._parser = RawConfigParser()
        self.filename = self.createFilename()
        if path_exists(self.filename):
            self._parser.read([self.filename])

        # Fusil application options
        self.fusil_max_memory = self.getint('fusil', 'max_memory', 100*1024*1024)
        self.fusil_success_score = self.getfloat('fusil', 'success_score', 0.50)
        self.fusil_error_score = self.getfloat('fusil', 'error_score', -0.50)
        self.fusil_success = self.getint('fusil', 'success', 1)
        self.fusil_session = self.getint('fusil', 'session', 0)
        self.fusil_normal_calm_load = self.getfloat('fusil', 'normal_calm_load', 0.50)
        self.fusil_normal_calm_sleep = self.getfloat('fusil', 'normal_calm_sleep', 0.5)
        self.fusil_slow_calm_load = self.getfloat('fusil', 'slow_calm_load', 0.30)
        self.fusil_slow_calm_sleep = self.getfloat('fusil', 'slow_calm_sleep', 3.0)
        self.fusil_xhost_program = self.getstr('fusil', 'xhost_program', 'xhost')

        # Process options
        self.use_cpu_probe = self.getbool('process', 'use_cpu_probe', True)
        self.process_max_memory = self.getint('process', 'max_memory', 100*1024*1024)
        self.process_core_dump = self.getbool('process', 'core_dump', True)
        self.process_max_user_process = self.getint('process', 'max_user_process', 10)

        # User used for subprocess
        self.process_user = self.getstr('process', 'user', 'fusil')
        self.process_uid = None

        # Group used for subprocess
        self.process_group = self.getstr('process', 'group', 'fusil')
        self.process_gid = None

        # Debugger options
        self.use_debugger = self.getbool('debugger', 'use_debugger', True)
        self.debugger_trace_forks = self.getbool('debugger', 'trace_forks', False)

        self._parser = None

    def createFilename(self):
        configdir = getenv("XDG_CONFIG_HOME")
        if not configdir:
            homedir = getenv("HOME")
            if not homedir:
                raise ConfigError("Unable to retreive user home directory: empty HOME environment variable")
            configdir = path_join(homedir, ".config")
        return path_join(configdir, "fusil.conf")

    def _gettype(self, func, type_name, section, key, default_value):
        try:
            value = func(section, key)
            if func == self._parser.get:
                value = value.strip()
            return value
        except (NoSectionError, NoOptionError):
            return default_value
        except ValueError, err:
            raise ConfigError("Value %s of section %s is not %s! %s" % (
                key, section, type_name, err))

    def getstr(self, section, key, default_value=None):
        return self._gettype(self._parser.get, "a string", section, key, default_value)

    def getbool(self, section, key, default_value):
        return self._gettype(self._parser.getboolean, "a boolean", section, key, default_value)

    def getint(self, section, key, default_value):
        return self._gettype(self._parser.getint, "an integer", section, key, default_value)

    def getfloat(self, section, key, default_value):
        return self._gettype(self._parser.getfloat, "a float", section, key, default_value)

